# Preprocessing splits of data

In [3]:
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split

In [7]:
#load clean dataset
data_path = Path(
    "Financial Phrasebank Project/data/processed/"
    "financial_phrasebank_66_clean.csv"
)

df_66_clean = pd.read_csv(data_path)
df_66_clean.shape
df_66_clean.head()


,sentence,label,sentence_clean
0,"According to Gran , the company has no plans t...",1,"According to Gran , the company has no plans t..."
1,Technopolis plans to develop in stages an area...,1,Technopolis plans to develop in stages an area...
2,With the new production plant the company woul...,2,With the new production plant the company woul...
3,According to the company 's updated strategy f...,2,According to the company 's updated strategy f...
4,"For the last quarter of 2010 , Componenta 's n...",2,"For the last quarter of 2010 , Componenta 's n..."


In [8]:
#produce the first split
#we will split the data into 70%train, 30%remainder
# which in this 30%, 15% will be validation and 15% test
train_df, temp_df = train_test_split(
    df_66_clean,
    test_size=0.30,
    random_state=42,
    stratify=df_66_clean["label"]
)

In [9]:
print("Full dataset:", len(df_66_clean))
print("Training set:", len(train_df))
print("Remaining set:", len(temp_df))

Full dataset: 4211
Training set: 2947
Remaining set: 1264


In [10]:
#Now we split the 
# 30% remainder into 15%validation and 15%test
val_df, test_df = train_test_split(
    temp_df,
    test_size=0.50,
    random_state=42,
    stratify=temp_df["label"]
)

In [11]:
print("Full dataset:", len(df_66_clean))
print("Training set:", len(train_df))
print("Validation set:", len(val_df))
print("Test set:", len(test_df))

Full dataset: 4211
Training set: 2947
Validation set: 632
Test set: 632


In [12]:
#check class balance in each of the splits
print("Training:")
print(train_df["label"].value_counts(normalize=True).sort_index())

print("\nValidation:")
print(val_df["label"].value_counts(normalize=True).sort_index())

print("\nTest:")
print(test_df["label"].value_counts(normalize=True).sort_index())

Training:
label
0    0.122158
1    0.600611
2    0.277231
Name: proportion, dtype: float64

Validation:
label
0    0.121835
1    0.599684
2    0.278481
Name: proportion, dtype: float64

Test:
label
0    0.121835
1    0.601266
2    0.276899
Name: proportion, dtype: float64


In [13]:
#leakage check 
train_sentences = set(train_df["sentence_clean"])
val_sentences = set(val_df["sentence_clean"])
test_sentences = set(test_df["sentence_clean"])

print("Train ↔ Validation overlap:",
      len(train_sentences & val_sentences))

print("Train ↔ Test overlap:",
      len(train_sentences & test_sentences))

print("Validation ↔ Test overlap:",
      len(val_sentences & test_sentences))

Train ↔ Validation overlap: 0
Train ↔ Test overlap: 0
Validation ↔ Test overlap: 0


In [14]:
#save the splits
train_df = train_df.reset_index(drop=True)
val_df = val_df.reset_index(drop=True)
test_df = test_df.reset_index(drop=True)

In [15]:
#create a splits folder and save the split data
from pathlib import Path

splits_dir = Path(
    "Financial Phrasebank Project/data/splits"
)

splits_dir.mkdir(
    parents=True,
    exist_ok=True
)

train_df.to_csv(
    splits_dir / "train.csv",
    index=False
)

val_df.to_csv(
    splits_dir / "validation.csv",
    index=False
)

test_df.to_csv(
    splits_dir / "test.csv",
    index=False
)

In [16]:
print("Train:", train_df.shape)
print("Validation:", val_df.shape)
print("Test:", test_df.shape)

Train: (2947, 3)
Validation: (632, 3)
Test: (632, 3)


## Split Summary

The cleaned dataset of 4,211 unique sentences was divided using a stratified
70/15/15 train-validation-test split with `random_state=42`.

- Training set: 2,947 observations
- Validation set: 632 observations
- Test set: 632 observations

Stratification preserved the original sentiment class distribution across all
three subsets.

No exact sentence overlap was found between the training, validation, and test
sets.

The test set will remain isolated during model development and will only be
used for final evaluation.